nil/services/journald_forwarder/clickhouse.go

```diff
@@ -0,0 +1,161 @@
+package journald_forwarder
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/ClickHouse/clickhouse-go/v2"
+	"github.com/ClickHouse/clickhouse-go/v2/lib/driver"
+	"github.com/NilFoundation/nil/nil/common/logging"
+)
+
+const (
+	DefaultDatabase = "nil"
+	DefaultTable    = "events"
+)
+
+type ClickhouseConfig struct {
+	ClickhouseAddr string
+	ListenAddr     string
+	DbUser         string
+	DbDatabase     string
+	DbPassword     string
+}
+
+type Clickhouse struct {
+	ClickhouseConfig
+	connect clickhouse.Conn
+}
+
+func (s *Clickhouse) InsertColumnsInTable(ctx context.Context,
+	database, tableName string,
+	columns, columnsType []string,
+) error {
+	columnClauses := make([]string, len(columns))
+	for i, column := range columns {
+		columnClauses[i] = fmt.Sprintf("ADD COLUMN %s %s", column, columnsType[i])
+	}
+
+	query := fmt.Sprintf(
+		"ALTER TABLE %s.%s %s;",
+		database,
+		tableName,
+		strings.Join(columnClauses, ", "),
+	)
+
+	return s.connect.Exec(ctx, query)
+}
+
+func (s *Clickhouse) InsertData(
+	ctx context.Context,
+	database string,
+	tableName string,
+	columns []string,
+	values [][]any,
+) error {
+	placeholders := make([]string, len(columns))
+	for i := range columns {
+		placeholders[i] = "?"
+	}
+
+	// Create multiple value placeholders for batch insert
+	valueRows := make([]string, len(values))
+	for i := range values {
+		valueRows[i] = fmt.Sprintf("(%s)", strings.Join(placeholders, ", "))
+	}
+
+	query := fmt.Sprintf(
+		"INSERT INTO %s.%s (%s) VALUES %s",
+		database,
+		tableName,
+		strings.Join(columns, ", "),
+		strings.Join(valueRows, ", "),
+	)
+
+	// Flatten values slice
+	flattenedValues := make([]any, 0, len(values)*len(columns))
+	for _, row := range values {
+		flattenedValues = append(flattenedValues, row...)
+	}
+
+	return s.connect.Exec(ctx, query, flattenedValues...)
+}
+
+func (s *Clickhouse) GetTabelColumnNames(ctx context.Context, database, tableName string) (map[string]any, error) {
+	const columnQuery = "SELECT name FROM system.columns WHERE database = ? AND table = ?"
+
+	rows, err := s.connect.Query(ctx, columnQuery, database, tableName)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	columns := make(map[string]any)
```

Comment:
Is a dictionary really necessary here? I understand that we should not have duplicates. A sorted array (for which the diff operation should also be performed in linear time) would suffice.

Reply:
Okay, I guess we also have unordered columns in the events, in which case it's really easier to do the check in the hash table than to add sorting there as well.